[READY] Add range support to completer command

micbou opened the pull request:
Implement the API introduced by PR https://github.com/Valloric/ycmd/pull/920.

PR https://github.com/Valloric/YouCompleteMe/pull/2898 is required for a bug-free formatting experience.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2900)
<!-- Reviewable:end -->

bstaletic commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[autoload/youcompleteme.vim, line 797 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L4t3SL9QEaJkPgTrw4z:-L4t3SL9QEaJkPgTrw5-:b9dua5k) ([raw file](https://github.com/valloric/youcompleteme/blob/dd9ca31d787b3a1508a82fc20aa7b450957adee7/autoload/youcompleteme.vim#L797)):*
> ```vim
>         \ YcmCompleter call s:CompleterCommand(0, <f-args>)
>   command! -nargs=* -complete=custom,youcompleteme#SubCommandsComplete -range=%
>         \ YcmCompleter call s:CompleterCommand(<range>, <f-args>) |
> ```

I believe that by doing this you've deleted the non-ranged command.
The `!` in `command!` says "if a command of the same name already exists, just replace it with this one."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[autoload/youcompleteme.vim, line 797 at r1](https://beta.reviewable.io/reviews/valloric/youcompleteme/2900#-L4t3SL9QEaJkPgTrw4z:-L4tNjC4CZ3t1WzObddR:bbdt60u) ([raw file](https://github.com/valloric/youcompleteme/blob/dd9ca31d787b3a1508a82fc20aa7b450957adee7/autoload/youcompleteme.vim#L797)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I believe that by doing this you've deleted the non-ranged command.
The `!` in `command!` says "if a command of the same name already exists, just replace it with this one."
</blockquote></details>

According to `:h :command-range`, `-range=%` means "allow range with the default being whole buffer". So we definitely don't need the non-ranged command any more.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[autoload/youcompleteme.vim, line 797 at r1](https://beta.reviewable.io/reviews/valloric/youcompleteme/2900#-L4t3SL9QEaJkPgTrw4z:-L4v1PUL88QdVdi1TeJ0:b-pdhsau) ([raw file](https://github.com/valloric/youcompleteme/blob/dd9ca31d787b3a1508a82fc20aa7b450957adee7/autoload/youcompleteme.vim#L797)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

According to `:h :command-range`, `-range=%` means "allow range with the default being whole buffer". So we definitely don't need the non-ranged command any more.
</blockquote></details>

You are right. I removed the non-ranged command.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

micbou commented:
Was it working? From my experience, `mode()` is always returning `n` when running a command even after a visual selection.

I didn't think of `vim.current.buffer.mark` to get the visual position. It's generally faster to use the Python API so I'll update the PR accordingly.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L5-ChmTQ6CxBXsj_MXA:b-twquqo)*
<!-- Sent from Reviewable.io -->

puremourning commented:
I think it was working, yeah. Then again, i note that the "_InVisualMode" call is _somewhat_ commented out, so that's probably the reason.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L5-De2Ouf69kD0Cy2ch:ba0q5w0)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Need to update the help to stay that the `:YcmCompleter` command can take a range. And probably to add the `Format` command.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[autoload/youcompleteme.vim, line 794 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-Fgp9G6nNlSZw9nFS:-L5-Fgp9G6nNlSZw9nFT:bv2betj) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/autoload/youcompleteme.vim#L794)):*
> ```vim
>   command! -nargs=* -complete=custom,youcompleteme#LogsComplete
>         \ YcmToggleLogs call s:ToggleLogs(<f-args>)
>   command! -nargs=* -complete=custom,youcompleteme#SubCommandsComplete -range=%
> ```

Defaulting the range to the whole file makes sense for formatting, but does it make sense for everything that we might use here? I guess this is a limitation of having an uber command `YcmCompleter` that does lots of stuff. 

Hmm.

For example, the other use-case I had was for code-actions (e.g. 'extract method' or 'organise imports'). We wouldn't want to apply that by default to the whole file range (probably more the current cursor position), but we still would want to use a real range for them.

It's tricky, so just thinking out loud really.

---

*[autoload/youcompleteme.vim, line 795 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-HBrZk9n-z05Ahrkj:-L5-HBrZk9n-z05Ahrkk:b-f1fy8p) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/autoload/youcompleteme.vim#L795)):*
> ```vim
>         \ YcmToggleLogs call s:ToggleLogs(<f-args>)
>   command! -nargs=* -complete=custom,youcompleteme#SubCommandsComplete -range=%
>         \ YcmCompleter call s:CompleterCommand(<range>, <f-args>)
> ```

If my reading of the vim doc is correct, this will always be nonzero?

Here's my justification, based purely on the text in the help:

This `<range>` expands as follows:

```
						*<range>*
	<range> The number of items in the command range: 0, 1 or 2
```

And `-range=%` defaults to the "whole file":

```
	-range=%    Range allowed, default is whole file (1,$)
```

Does that actually marry with your testing? or am i reading it wrong?

TBH the description of `<range>` is kind of vague. Why would there only be 0,1 or 2 elements?

---

*[python/ycm/vimsupport.py, line 1060 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-IxbYUBdbGnyyYZO3:-L5-IxbYUBdbGnyyYZO4:br7arm0) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/vimsupport.py#L1060)):*
> ```Python
> 
> 
> def GetVisualRange():
> ```

my comments on this method are: take a look at my version. take anything you like and discard anything else :)

personally i think the comment about the weird addressing is useful to repeat, and also using `vim.current.buffer.mark`

---

*[python/ycm/youcompleteme.py, line 325 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-Jbuvz2_CON95cA92:-L5-Jbuvz2_CON95cA93:bc5jzex) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/youcompleteme.py#L325)):*
> ```Python
>       }
>     }
>     if visual_mode:
> ```

are the `<` and `>` marks set when _not_ in visual mode?

If i'm reading this right, you can _only_ use a range from visual mode. I think if we're going to allow a range on the `:YcmCompleter` command, we should consider other ways of specifying a range such as `:.,+5YcmCompleter Format`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L5-K8TeVM3B1beyPw-a:bsfud9e)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[python/ycm/tests/test_utils.py, line 206 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-KQ5qsaCg93SKTlsU:-L5-KQ5qsaCg93SKTlsV:b6gr1hh) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/tests/test_utils.py#L206)):*
> ```Python
> 
>   if value == 'shiftwidth()':
>     return 4
> ```

criminal! shift widths should be 2 spaces in YCM land! :D

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Reviewed 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 794 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-Fgp9G6nNlSZw9nFS:-L5-rCsDHeK7nVK4dvet:bg6ra8i) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/autoload/youcompleteme.vim#L794)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Defaulting the range to the whole file makes sense for formatting, but does it make sense for everything that we might use here? I guess this is a limitation of having an uber command `YcmCompleter` that does lots of stuff. 

Hmm.

For example, the other use-case I had was for code-actions (e.g. 'extract method' or 'organise imports'). We wouldn't want to apply that by default to the whole file range (probably more the current cursor position), but we still would want to use a real range for them.

It's tricky, so just thinking out loud really.
</blockquote></details>

We could default it to a single line. Formatting would act just like `==` or `gqgq`. At least we'd be consistent with what vim provides.

---

*[autoload/youcompleteme.vim, line 795 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-HBrZk9n-z05Ahrkj:-L5-ruMgdctRsXtAsRFn:bqqv1pk) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/autoload/youcompleteme.vim#L795)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

If my reading of the vim doc is correct, this will always be nonzero?

Here's my justification, based purely on the text in the help:

This `<range>` expands as follows:

```
						*<range>*
	<range> The number of items in the command range: 0, 1 or 2
```

And `-range=%` defaults to the "whole file":

```
	-range=%    Range allowed, default is whole file (1,$)
```

Does that actually marry with your testing? or am i reading it wrong?

TBH the description of `<range>` is kind of vague. Why would there only be 0,1 or 2 elements?


</blockquote></details>

If I'm reading the help correctly:

- `<range>` is 0, meaning no range is given (`:YcmCompleter Format`). Here, whatever is default is used for the range.
- `<range>` is 1, meaning only one number is provided for the range (`:23d 4`). Here it means "do the command on the specified line.
- `<range>` is 2, meaning the full range is specified (`:5,6t$`). It means "consider this range of lines as the input".

In case of no range specified, range is determined by `-range`.

- `-rage=%` - whole file.
- `-range` - current line.
- `-range=N` - a count, not a range. Default is `N` and is specified before the command name `:5d`.
- `-count=N` - a count. Can be specified as a prefix or sufix. `:4d` or `:Next5`.

---

*[python/ycm/vimsupport.py, line 1062 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-zB4o4uQYpPsmuW_B:-L5-zB4pVG3F3xkf91jH:b-6rqnem) ([raw file](https://github.com/valloric/youcompleteme/blob/83087f265b5dabaab1018aec1ed07bda153d4f1a/python/ycm/vimsupport.py#L1062)):*
> ```Python
> def GetVisualRange():
>   start = vim.current.buffer.mark( '<' )
>   end = vim.current.buffer.mark( '>' )
> ```

Tests say `AttributeError: 'VimBuffer' object has no attribute 'mark'`.

---

*[python/ycm/youcompleteme.py, line 325 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-Jbuvz2_CON95cA92:-L5-vzBh43JSKrsG3_g4:b-g19hom) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/youcompleteme.py#L325)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

are the `<` and `>` marks set when _not_ in visual mode?

If i'm reading this right, you can _only_ use a range from visual mode. I think if we're going to allow a range on the `:YcmCompleter` command, we should consider other ways of specifying a range such as `:.,+5YcmCompleter Format`
</blockquote></details>

 The marks are remembered from the last visual selection in the buffer. Marks are not preserved over restarts by default.

---

*[python/ycm/tests/test_utils.py, line 206 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-KQ5qsaCg93SKTlsU:-L5-wijEr9_q8rbPL6kg:bs3wjzt) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/tests/test_utils.py#L206)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

criminal! shift widths should be 2 spaces in YCM land! :D
</blockquote></details>

8 spaces deep tab characters, as far as Linux devlopment and Dennis Ritchie are concerned.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Now we need to document in vim's help that YcmCompleter takes range.

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 794 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-Fgp9G6nNlSZw9nFS:-L52fcpISMlfg_th4q7v:bz7ar5q) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/autoload/youcompleteme.vim#L794)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

We could default it to a single line. Formatting would act just like `==` or `gqgq`. At least we'd be consistent with what vim provides.
</blockquote></details>

I really do thing we should default tojust the current line, not the whole file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L52fLNGdMK6EG6NfMqL:bzdtq9z)*
<!-- Sent from Reviewable.io -->

micbou commented:
Updated the README.

---

Reviewed 1 of 2 files at r3, 4 of 5 files at r4, 2 of 2 files at r5, 3 of 3 files at r6, 2 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[autoload/youcompleteme.vim, line 794 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-Fgp9G6nNlSZw9nFS:-L54_HGkag85Tln_jfMd:b-7wmn6l) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/autoload/youcompleteme.vim#L794)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I really do thing we should default tojust the current line, not the whole file.
</blockquote></details>

If no range is given, `<range>` returns 0 and thus we don't add the `range` field to the command request. In other words, the behavior of a command when no range is given depends on the implementation of that command in ycmd. So, I think we are fine. Still, there is no point to use `-range=%` over `-range` so we are using the latter now.

---

*[autoload/youcompleteme.vim, line 795 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-HBrZk9n-z05Ahrkj:-L54b_LpXpszQTx0Ln0S:b-111d4d) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/autoload/youcompleteme.vim#L795)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

If I'm reading the help correctly:

- `<range>` is 0, meaning no range is given (`:YcmCompleter Format`). Here, whatever is default is used for the range.
- `<range>` is 1, meaning only one number is provided for the range (`:23d 4`). Here it means "do the command on the specified line.
- `<range>` is 2, meaning the full range is specified (`:5,6t$`). It means "consider this range of lines as the input".

In case of no range specified, range is determined by `-range`.

- `-rage=%` - whole file.
- `-range` - current line.
- `-range=N` - a count, not a range. Default is `N` and is specified before the command name `:5d`.
- `-count=N` - a count. Can be specified as a prefix or sufix. `:4d` or `:Next5`.
</blockquote></details>

What @bstaletic said.

---

*[python/ycm/vimsupport.py, line 1060 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-IxbYUBdbGnyyYZO3:-L54_C6_B6CY-1ef_MPt:b6brv9q) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/vimsupport.py#L1060)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

my comments on this method are: take a look at my version. take anything you like and discard anything else :)

personally i think the comment about the weird addressing is useful to repeat, and also using `vim.current.buffer.mark`
</blockquote></details>

I took almost everything. Thanks!

---

*[python/ycm/vimsupport.py, line 1062 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-zB4o4uQYpPsmuW_B:-L51fP3OqL-8RkoQgwSh:bt2vr1b) ([raw file](https://github.com/valloric/youcompleteme/blob/83087f265b5dabaab1018aec1ed07bda153d4f1a/python/ycm/vimsupport.py#L1062)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Tests say `AttributeError: 'VimBuffer' object has no attribute 'mark'`.
</blockquote></details>

Fixed.

---

*[python/ycm/youcompleteme.py, line 325 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-Jbuvz2_CON95cA92:-L54YKoRMm-fOdUFwhUB:bx6f1ur) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/youcompleteme.py#L325)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

 The marks are remembered from the last visual selection in the buffer. Marks are not preserved over restarts by default.
</blockquote></details>

I didn't think of that. I've updated the PR to properly behave when a range is given on the command line. Since only the lines of the range can be specified this way (the `<line1>` and `<line2>` variables), we compare them to the last visual selection and if they coincide, we use the columns of that selection. It's a little hacky but should work well in practice.

---

*[python/ycm/tests/test_utils.py, line 206 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L5-KQ5qsaCg93SKTlsU:-L54Y6FCvNfj6cB98PGr:b-yn4xii) ([raw file](https://github.com/valloric/youcompleteme/blob/515567db9654d1f234de3aea10ae401308b6dc23/python/ycm/tests/test_utils.py#L206)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

8 spaces deep tab characters, as far as Linux devlopment and Dennis Ritchie are concerned.
</blockquote></details>

Fixed because it's important stuff.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L54YtSA8VkUBuOJIX8T:b-q4480p)*
<!-- Sent from Reviewable.io -->

micbou commented:
I'll update ycmd submodule once PR https://github.com/Valloric/ycmd/pull/920 has landed.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L54iv6LwPWq_dz4a8Vj:bb3ygv8)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Noice!

It works well.

The only one concern is that the quick fix list opening is kind of jarring because:
- for a large file, it like takes up the whole screen
- all of the entries are whitespace, so don't _really_ add much value

We should consider if we can suppress it (e.g. using silent mode or something similar). That can be a separate PR though if you like.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[README.md, line 1790 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L54iKzHE_eIkVuAT0jg:-L54iKzIu1iG6ndhiwnY:b-7ze0xr) ([raw file](https://github.com/valloric/youcompleteme/blob/e9aecda7a22f23e36f1a91cffb8a89092522b6ab/README.md#L1790)):*
> ```Markdown
> of the Vim options `shiftwidth` and `expandtab` (see `:h 'sw'` and `:h et`
> respectively). To format a specific part of your document, you can either select
> it in one of Vim visual modes (see `:h visual-use`) and run the command or
> ```

Vim's

---

*[doc/youcompleteme.txt, line 203 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L54ioM2kWaBIf4ED_X6:-L54ioM3qLztVScziPlh:bz1e6a5) ([raw file](https://github.com/valloric/youcompleteme/blob/c2a7b3cffcc4e7a9a2b0007a3ed6b75f039f06c5/doc/youcompleteme.txt#L203)):*
> ```
> 
> First carefully read the installation instructions for your OS. We recommend
> you use the supplied |install.py|.
> ```

so i think i was the cause of this changing before. are you by any chance using a different version of html2vimdoc from me? Or perhaps my command line is wrong. I was doing:

```
$ html2vimdoc -f youcompleteme README.md > doc/youcompleteme.txt
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L54jhrS-A8ueLlMRm3t:b-tz0x7q)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 2 of 5 files at r4, 2 of 3 files at r6, 2 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[README.md, line 1790 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L54iKzHE_eIkVuAT0jg:-L54oXMGQ15NaR94MFL7:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/e9aecda7a22f23e36f1a91cffb8a89092522b6ab/README.md#L1790)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Vim's
</blockquote></details>

Done.

---

*[doc/youcompleteme.txt, line 203 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L54ioM2kWaBIf4ED_X6:-L54nIU5FjquR_sBoZm5:b-qcgjzg) ([raw file](https://github.com/valloric/youcompleteme/blob/c2a7b3cffcc4e7a9a2b0007a3ed6b75f039f06c5/doc/youcompleteme.txt#L203)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

so i think i was the cause of this changing before. are you by any chance using a different version of html2vimdoc from me? Or perhaps my command line is wrong. I was doing:

```
$ html2vimdoc -f youcompleteme README.md > doc/youcompleteme.txt
```
</blockquote></details>

Yes, I am using a custom version (available [here](https://github.com/micbou/vim-tools/tree/vim-tags)) that fixes issue https://github.com/Valloric/YouCompleteMe/issues/2479.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Reviewed 2 of 2 files at r5, 2 of 3 files at r6, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[python/ycm/tests/command_test.py, line 82 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L550SwtnPEsZ9csB4bS:-L550SwtnPEsZ9csB4bT:b-qtf2q6) ([raw file](https://github.com/valloric/youcompleteme/blob/74d675d68807a14f65378c2f90275c987303f00f/python/ycm/tests/command_test.py#L82)):*
> ```Python
> 
> @YouCompleteMeInstance()
> def SendCommandRequest_BuildRange_NoVisualMarks_test( ycm, *args ):
> ```

This test timeouted on travis on python2. I restarted the build.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

micbou commented:
Updated the PR with the latest version of ycmd and changed the `Format` command to not populate and open the quickfix window. PR should be ready.

---

Reviewed 1 of 8 files at r10, 9 of 9 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L56OZJ9dUTDlUEZCsad:bxu8u9u)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/client/command_request.py, line 119 at r11](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L56Rpcx1SankIiX-MIN:-L56Rpcx1SankIiX-MIO:b-x2x5jt) ([raw file](https://github.com/valloric/youcompleteme/blob/819c61b1ec5be89f9aaacb4433231612d95c5ac8/python/ycm/client/command_request.py#L119)):*
> ```Python
>         vimsupport.ReplaceChunks(
>           self._response[ 'fixits' ][ fixit_index ][ 'chunks' ],
>           self._command == 'Format' )
> ```

I would use `silent = self._command == 'Format'` for ease of reading. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L56RziTAizAkqKH67Nu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 1 of 8 files at r10, 9 of 9 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 2 files at r12.
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[python/ycm/client/command_request.py, line 119 at r11](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-L56Rpcx1SankIiX-MIN:-L56TBPW2kcT9gTYQjud:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/819c61b1ec5be89f9aaacb4433231612d95c5ac8/python/ycm/client/command_request.py#L119)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

I would use `silent = self._command == 'Format'` for ease of reading. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
@zzbot r=puremourning

---

Reviewed 2 of 2 files at r9, 1 of 8 files at r10, 9 of 9 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2900#-:-L56ToEr3y2u1OPZEdu0:bsrfr92)*
<!-- Sent from Reviewable.io -->
